### PR TITLE
[Refactor] Basic content for I18n on Admin panel

### PR DESCRIPTION
**Pull request by aitbw**
References
===================
* **Related PR**: #1580

Objectives
===================
## Added
* `I18nContent` model specs
* Tests and examples for `flat_hash` function (by @raul-fuentes)

## Fixes
* Fixed a bug where all matching I18n keys wouldn't appear when the `#begins_with_key` scope was used because of a trailing question mark (`?`)

## Refactoring (for DRYness/readability)
* `globalize_locales` partials to use lambdas and extracted shared markup into a partial (by @javierm)
* Extracted variable assignment from `information_texts/_form_field` partial into helper method
* Avoid using ternary operator when appending/creating I18n keys
* Concise variable names for `I18nContent#flat_hash` method

Visual Changes
===================
* None, all stays the same

Notes
===================
* Re: `InformationTexts` controller:
  * Refactor still pending considering its complexity (if possible)
  * Unit tests must be written

* Backport to CONSUL once approved and merged into `master`

### CODE REVIEW

File: spec/models/i18n_content_spec.rb [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+      expect(I18nContent.all.size).to eq(2)
+
+      query = I18nContent.by_key(key)
+
+      expect(query.size).to eq(1)
+      expect(query).to eq([debate_title])
+    end
+
+    it 'return all matching records when #begins_with_key is used' do
+      debate_translation = create(:i18n_content)
```
Comment (javierm) [inline spec/models/i18n_content_spec.rb]: A very small thing: I think the test is easier to understand if we don't rely on factory defaults (which by the way could change in the future for whatever reason). Something like `debate_translation = create(:18n_content, key: "debates.title")` makes it easier to understand why `I18nContent.begins_with_key('debates')` should return two records.

File: app/controllers/admin/site_customization/information_texts_controller.rb [hunk 2]
```diff
@@ -13,7 +13,7 @@ def update
 
       unless values.empty?
         values.each do |key, value|
-          locale = key.split("_").last
+          locale = key.split('_').last
```
[the remark this answers is not in the document]
Comment (aitbw) [reply app/controllers/admin/site_customization/information_texts_controller.rb]: We don't but I like to stay consistent regarding when it comes to single/double quotes: single quotes unless interpolation is necessary on backend and double quotes without exception on frontend

File: app/controllers/admin/site_customization/information_texts_controller.rb [hunk 3]
```diff
@@ -74,17 +81,11 @@ def append_or_create_keys
       translations = I18n.backend.send(:translations)[locale.to_sym]
 
       translations.each do |k, v|
-        @content[k.to_s] = flat_hash(v).keys
-                                       .map { |s| @existing_keys["#{k.to_s}.#{s}"].nil? ?
-                                              I18nContent.new(key: "#{k.to_s}.#{s}") :
-                                              @existing_keys["#{k.to_s}.#{s}"] }
+        @content[k.to_s] = I18nContent.flat_hash(v).keys.map { |s|
+          @existing_keys["#{k.to_s}.#{s}"].nil? ? I18nContent.new(key: "#{k.to_s}.#{s}") :
+                                                  @existing_keys["#{k.to_s}.#{s}"]
```
[the remark this answers is not in the document]
Comment (aitbw) [reply app/controllers/admin/site_customization/information_texts_controller.rb]: :thinking: yes, I think so too but first, we would need some tests for that method so we can rest assured it works/behaves as expected before **and** after the refactor

File: app/helpers/site_customization_helper.rb [hunk 4]
```diff
@@ -1,5 +1,22 @@
...
+    if i18n_content.present?
+      translation = I18nContentTranslation.where(
+        i18n_content_id: i18n_content.id,
+        locale: locale
+      ).first.try(:value)
+    else
+      translation = false
+    end
+
+    return translation
```
[the remark this answers is not in the document]
Comment (aitbw) [reply app/helpers/site_customization_helper.rb]: You're right @javierm, thanks! :tada:

File: spec/models/i18n_content_spec.rb (hunk 1 above)
Comment (aitbw) [reply spec/models/i18n_content_spec.rb]: You're right, @javierm I must have missed that one, thanks! :heart:

File: app/controllers/admin/site_customization/information_texts_controller.rb (hunk 3 above)
[the remark this answers is not in the document]
Comment (javierm) [reply app/controllers/admin/site_customization/information_texts_controller.rb]: You're absolutely right. Either leave it as it is or add tests and refactor; up to you!

**Comment by javierm** [comment]
@aitbw While reviewing another `Translatable` related issue I realized the files in `app/views/admin/shared/_globalize_locales.html.erb` and `app/views/admin/site_customization/information_texts/_globalize_locales.html.erb` are almost identical.

I've done a quick patch to avoid duplication in these files and have changed the methods ending with `_display_translation?` a bit to make them a bit easier to understand (or so I think :sweat_smile:).

I'm not sure if it's reasonable to add it (or similar changes) to this PR :thinking:. What do you think? Is it worth it? Is it an unneeded refactoring?

Here's the patch (as you might see, it's just a draft):

```diff
diff --git a/app/helpers/globalize_helper.rb b/app/helpers/globalize_helper.rb
index b64c1ff..9b54bee 100644
--- a/app/helpers/globalize_helper.rb
+++ b/app/helpers/globalize_helper.rb
@@ -14,8 +14,8 @@ module GlobalizeHelper
     same_locale?(neutral_locale(I18n.locale), neutral_locale(locale)) ? "" : "display: none"
   end
 
-  def css_to_display_translation?(resource, locale)
-    resource.translated_locales.include?(neutral_locale(locale)) || locale == I18n.locale ? "" : "display: none"
+  def display_translation_for_resource?(resource, locale)
+    resource.translated_locales.include?(neutral_locale(locale)) || locale == I18n.locale
   end
 
   def highlight_current?(locale)
diff --git a/app/helpers/site_customization_helper.rb b/app/helpers/site_customization_helper.rb
index e81d5d8..fadb5e2 100644
--- a/app/helpers/site_customization_helper.rb
+++ b/app/helpers/site_customization_helper.rb
@@ -1,7 +1,7 @@
 module SiteCustomizationHelper
 
-  def site_customization_display_translation?(locale)
-    I18nContentTranslation.existing_languages.include?(neutral_locale(locale)) || locale == I18n.locale ? "" : "display: none"
+  def site_customization_enable_translation?(locale)
+    I18nContentTranslation.existing_languages.include?(neutral_locale(locale)) || locale == I18n.locale
   end
 
   def translation_for_locale(content, locale)
diff --git a/app/views/admin/shared/_common_globalize_locales.html.erb b/app/views/admin/shared/_common_globalize_locales.html.erb
new file mode 100644
index 0000000..5f770dc
--- /dev/null
+++ b/app/views/admin/shared/_common_globalize_locales.html.erb
@@ -0,0 +1,27 @@
+<% I18n.available_locales.each do |locale| %>
+  <%= link_to t("admin.translations.remove_language"), "#",
+              id: "delete-#{neutral_locale(locale)}",
+              style: show_delete?(locale),
+              class: 'float-right delete js-delete-language',
+              data: { locale: neutral_locale(locale) } %>
+
+<% end %>
+
+<ul class="tabs" data-tabs id="globalize_locale">
+  <% I18n.available_locales.each do |locale| %>
+    <li class="tabs-title">
+      <%= link_to name_for_locale(locale), "#",
+                  style: display_locale_condition.call(locale) ? "" : "display: none",
+                  class: "js-globalize-locale-link #{highlight_current?(locale)}",
+                  data: { locale: neutral_locale(locale) },
+                  remote: true %>
+    </li>
+  <% end %>
+</ul>
+
+<div class="small-12 medium-6">
+  <%= select_tag :translation_locale,
+                 options_for_locale_select,
+                 prompt: t("admin.translations.add_language"),
+                 class: "js-globalize-locale" %>
+</div>
diff --git a/app/views/admin/shared/_globalize_locales.html.erb b/app/views/admin/shared/_globalize_locales.html.erb
index 977fa6d..db5194a 100644
--- a/app/views/admin/shared/_globalize_locales.html.erb
+++ b/app/views/admin/shared/_globalize_locales.html.erb
@@ -1,27 +1,2 @@
-<% I18n.available_locales.each do |locale| %>
-  <%= link_to t("admin.translations.remove_language"), "#",
-              id: "delete-#{neutral_locale(locale)}",
-              style: show_delete?(locale),
-              class: 'float-right delete js-delete-language',
-              data: { locale: neutral_locale(locale) } %>
-
-<% end %>
-
-<ul class="tabs" data-tabs id="globalize_locale">
-  <% I18n.available_locales.each do |locale| %>
-    <li class="tabs-title">
-      <%= link_to name_for_locale(locale), "#",
-                  style: css_to_display_translation?(resource, locale),
-                  class: "js-globalize-locale-link #{highlight_current?(locale)}",
-                  data: { locale: neutral_locale(locale) },
-                  remote: true %>
-    </li>
-  <% end %>
-</ul>
-
-<div class="small-12 medium-6">
-  <%= select_tag :translation_locale,
-                 options_for_locale_select,
-                 prompt: t("admin.translations.add_language"),
-                 class: "js-globalize-locale" %>
-</div>
+<%= render "admin/shared/common_globalize_locales",
+  display_locale_condition: lambda { |locale| display_translation_for_resource?(resource, locale) } %>
diff --git a/app/views/admin/site_customization/information_texts/_globalize_locales.html.erb b/app/views/admin/site_customization/information_texts/_globalize_locales.html.erb
index f69e488..78994f5 100644
--- a/app/views/admin/site_customization/information_texts/_globalize_locales.html.erb
+++ b/app/views/admin/site_customization/information_texts/_globalize_locales.html.erb
@@ -1,27 +1 @@
-<% I18n.available_locales.each do |locale| %>
-  <%= link_to t("admin.translations.remove_language"), "#",
-              id: "delete-#{neutral_locale(locale)}",
-              style: show_delete?(locale),
-              class: 'float-right delete js-delete-language',
-              data: { locale: neutral_locale(locale) } %>
-
-<% end %>
-
-<ul class="tabs" data-tabs id="globalize_locale">
-  <% I18n.available_locales.each do |locale| %>
-    <li class="tabs-title">
-      <%= link_to name_for_locale(locale), "#",
-                  style: site_customization_display_translation?(locale),
-                  class: "js-globalize-locale-link #{highlight_current?(locale)}",
-                  data: { locale: neutral_locale(locale) },
-                  remote: true %>
-    </li>
-  <% end %>
-</ul>
-
-<div class="small-12 medium-6">
-  <%= select_tag :translation_locale,
-                 options_for_locale_select,
-                 prompt: t("admin.translations.add_language"),
-                 class: "js-globalize-locale" %>
-</div>
+<%= render "admin/shared/common_globalize_locales", display_locale_condition: lambda { |locale| site_customization_enable_translation?(locale) } %>
```

Variable/method names can certainly be improved, and passing a `lambda` object to a partial isn't that great, but I'd rather do so than having two identical partials around. Ideas are welcome!

**Comment by aitbw** [comment]
Hey @javierm, thanks for the suggestion! I just pushed two new commits: one with your patch as it increases code DRYness (which is always a good thing) and I went a little bit further and fused the two helper methods for displaying translations into one (as they're basically the same) taking advantage of default params in order to reduce verbosity even further :smile_cat: what do you think?

File: app/helpers/globalize_helper.rb [hunk 5]
```diff
@@ -14,8 +14,12 @@ def display_translation?(locale)
...
   end
 
-  def display_translation_for_resource?(resource, locale)
-    resource.translated_locales.include?(neutral_locale(locale)) || locale == I18n.locale
+  def display_translation_for_resource?(locale, resource = nil)
+    if resource.nil?
+      I18nContentTranslation.existing_languages.include?(neutral_locale(locale)) || locale == I18n.locale
+    else
+      resource.translated_locales.include?(neutral_locale(locale)) || locale == I18n.locale
+    end
```
Comment (javierm) [inline app/helpers/globalize_helper.rb]: I think we usually avoid the `if something.nil? (...) else (...)` structure, and instead do something like:
```ruby
if resource
  resource.translated_locales.include?(neutral_locale(locale)) || locale == I18n.locale
else
  I18nContentTranslation.existing_languages.include?(neutral_locale(locale)) || locale == I18n.locale
end
```
Correct me if I'm wrong!

**Comment by javierm** [comment]
@aitbw The refactoring looks good! The only thing we'll have to look out for is we might have conflicts with #1602 and #1613 since they also deal with the same methods.

File: app/helpers/globalize_helper.rb (hunk 5 above)
Comment (aitbw) [reply app/helpers/globalize_helper.rb]: I did thought about that but I think `nil?` usage makes it more explicit and therefore, easier to read

**Comment by javierm** [comment]
The commits in this pull request were included in #1669. Thanks so much @aitbw!

**Pull request closed without merging**